### PR TITLE
API: replace() should raise an exception if invalid argument is given

### LINKED ISSUE #18634: Series.replace() should raise an exception if invalid argument is given
Series.replace() should potentially raise an exception if an invalid argument is given? I am new to pandas and this is more of a question. **This is a user error on my part, but possible improvement to replace.**

```python
df = pd.DataFrame(
{
    'one': ['1','1 ','10'],
    'two': ['1 ', '20 ', '30 ']
})

# creates
     one  two
0    1    1
1    1    20
2    10    30

# I intentionally added a space in df.one. So df.one.value_counts() results in
1     1
10    1
1     1

# I want to strip the spaces around all values in df.one 
# so that value_counts() only has 1 and 10 with values 2 and 1 respectively. 
# The following does not work.

df.one.replace(lambda x: x.strip(), inplace=True)

```
#### Problem description

Series.replace() accepts `"str, regex, list, dict, Series, numeric, or None"`. So I understand why the above does not work. But as @toobaz pointed out on gitter, when a lambda is passed, replace should probably raise an error?

#### Expected Output
Raises TypeError

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.3
pytest: None
pip: 9.0.1
setuptools: 36.3.0
Cython: None
numpy: 1.13.1
scipy: 0.19.1
xarray: None
IPython: 6.1.0
sphinx: 1.6.3
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.8
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: 3.8.0
bs4: None
html5lib: 0.999999999
sqlalchemy: 1.1.14
pymysql: 0.7.11.None
psycopg2: None
jinja2: 2.8.1
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by a-y-khan**
- [x] closes #18634
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I used `is_scalar()` instead of `is_numeric()` to allow data structures like pd.Timestamp to be used in `replace()`. See tests in pandas/tests/frame/methods/test_replace.py, pandas/tests/series/methods/test_replace.py for examples.

### CODE REVIEW

File: pandas/tests/frame/methods/test_replace.py [hunk 1]
```diff
@@ -1363,3 +1363,27 @@ def test_replace_after_convert_dtypes(self):
         result = df.replace(1, 10)
         expected = pd.DataFrame({"grp": [10, 2, 3, 4, 5]}, dtype="Int64")
         tm.assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "df, to_replace",
+        [
+            (
+                {"one": ["a", "b ", "c"], "two": ["d ", "e ", "f "]},
```
Comment (WillAyd) [inline pandas/tests/frame/methods/test_replace.py]: Since `df` is the same all of the time no need to parametrize; just place in the body

File: pandas/tests/frame/methods/test_replace.py [hunk 2]
```diff
@@ -1363,3 +1363,27 @@ def test_replace_after_convert_dtypes(self):
...
+                lambda x: x.strip(),
+            ),
+            (
+                {"one": ["a", "b ", "c"], "two": ["d ", "e ", "f "]},
+                type("callable_object", (object,), dict(__call__=lambda x: x.strip())),
```
Comment (WillAyd) [inline pandas/tests/frame/methods/test_replace.py]: This is practically the same as the lambda right? I think just the former is fine

File: pandas/tests/series/methods/test_replace.py [hunk 3]
```diff
@@ -362,3 +362,24 @@ def test_replace_no_cast(self, ser, exp):
...
+        "ser, to_replace",
+        [
+            (["a", "b", "c "], lambda x: x.strip()),
+            (
+                ["a", "b", "c "],
+                type("callable_object", (object,), dict(__call__=lambda x: x.strip())),
+            ),
+        ],
+    )
+    def test_replace_invalid_to_replace(self, ser, to_replace):
```
Comment (WillAyd) [inline pandas/tests/series/methods/test_replace.py]: Same comments here as on other test

File: pandas/core/generic.py [hunk 4]
```diff
@@ -6393,6 +6398,19 @@ def replace(
         regex=False,
         method="pad",
     ):
+        if not (
+            is_scalar(to_replace)
+            or isinstance(to_replace, pd.Series)
+            or is_re_compilable(to_replace)
+            or is_list_like(to_replace)
+            or is_dict_like(to_replace)
```
Comment (WillAyd) [inline pandas/core/generic.py]: `is_list_like` actually will cover the dict case, so don't need this line

File: pandas/core/generic.py [hunk 5]
```diff
@@ -6393,6 +6398,19 @@ def replace(
...
+        ):
+            raise TypeError(
+                f"Expecting 'to_replace' to be str, regex, list, dict, Series, "
```
Comment (WillAyd) [inline pandas/core/generic.py]:
```suggestion
                "Expecting 'to_replace' to be str, regex, list, dict, Series, "
```

Minor but this isn't necessary

File: pandas/core/generic.py [hunk 6]
```diff
@@ -6393,6 +6398,19 @@ def replace(
...
+                f"int, float or None, got invalid type "
```
Comment (WillAyd) [inline pandas/core/generic.py]:
```suggestion
                "int, float or None, got invalid type "
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for the PR

File: pandas/tests/frame/methods/test_replace.py (hunk 2 above)
Comment (a-y-khan) [reply pandas/tests/frame/methods/test_replace.py]: Yes, it is. I can just move the df and lambda to the body.

File: pandas/core/generic.py (hunk 4 above)
Comment (a-y-khan) [reply pandas/core/generic.py]: Removed, and thank you for the PR feedback!

**Review by WillAyd — APPROVED** [review]
lgtm

File: pandas/core/generic.py [hunk 7]
```diff
@@ -6172,7 +6172,11 @@ def bfill(
         AssertionError
             * If `regex` is not a ``bool`` and `to_replace` is not
               ``None``.
+
         TypeError
+            * If `to_replace` is not any of the expected types
```
Comment (jreback) [inline pandas/core/generic.py]: don't list the scalars, just say

scalar, array-like, dict or None

File: pandas/core/generic.py [hunk 8]
```diff
@@ -6376,6 +6381,18 @@ def replace(
...
         method="pad",
     ):
+        if not (
+            is_scalar(to_replace)
+            or isinstance(to_replace, pd.Series)
+            or is_re_compilable(to_replace)
+            or is_list_like(to_replace)
+        ):
+            raise TypeError(
+                "Expecting 'to_replace' to be str, regex, list, dict, Series, "
```
Comment (jreback) [inline pandas/core/generic.py]: reword this as well

File: doc/source/whatsnew/v1.1.0.rst [hunk 9]
```diff
@@ -57,6 +57,7 @@ Other API changes
   will now be ``min`` and ``max`` to match with numeric dtypes in :meth:`DataFrame.describe` (:issue:`30164`)
 - Added :meth:`DataFrame.value_counts` (:issue:`5377`)
 - :meth:`Groupby.groups` now returns an abbreviated representation when called on large dataframes (:issue:`1135`)
+- :meth:`DataFrame.replace` and :meth:`Series.replace` will raise a ``TypeError`` if ``to_replace`` is not an expected type. Previously the ``replace`` would fail silently (:issue:`18634`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: you can put this in bug fixes in reshaping section

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 7 above)
Comment (a-y-khan) [reply pandas/core/generic.py]: Done!

File: pandas/core/generic.py (hunk 8 above)
Comment (a-y-khan) [reply pandas/core/generic.py]: Also, reworded.

File: doc/source/whatsnew/v1.1.0.rst (hunk 9 above)
Comment (a-y-khan) [reply doc/source/whatsnew/v1.1.0.rst]: Done!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @a-y-khan

**Comment by a-y-khan** [comment]
Glad I could make a contribution!
